src/errum_cpp_routine.cpp

```diff
@@ -0,0 +1,564 @@
+    double aforw2 = .0;
+    double bforw2 = .0;
+    //double afornu = .0;
+    //update deltas and gammas across variables
+    //update intercepts first
+    arma::mat Ytilde(J,K);
+    if(V==1){
+        Ytilde = Y;
+    }
+    else{
+        Ytilde = Y - X.cols(1,V-1)*(Gamma.rows(1,V-1));
+    }
+    //arma::mat Ytilde = Y - X.cols(1,V-1)*(Gamma.rows(1,V-1));
+    double sigv2 = 1./(double(J)+1.);
+    arma::rowvec mu_v = sigv2*(arma::sum(Ytilde));
+    for(unsigned int k=0;k<K;++k){
+        Gamma(0,k) = R::rnorm(mu_v(k),sqrt(sigv2));
+    }
+    if(V>1){
+        for(unsigned int vv=1;vv<V;++vv){
+            arma::rowvec delta_v=deltas.row(vv);
+            arma::uvec notv = find(onetoV != vv);
+            arma::vec Xv = X.col(vv);
+            arma::mat Ytilde = Y - X.cols(notv)*(Gamma.rows(notv));
+            double sigv2 = 1./(diagXpX(vv)+w2);
+            arma::vec mu_v = sigv2*(Ytilde.t()*Xv);
+            arma::vec lnb_m_lna=log(1.-nu)-.5*pow(mu_v,2)/sigv2-log(nu)-.5*(log(w2)+log(sigv2));
+            for(unsigned int k=0;k<K;++k){
+                arma::uvec notk = find(onetoK != k);
+                double uvk = R::runif(0,1);
+                double lnuvk = log(uvk);
+                double ln1muvk = log(1.-uvk);
+                //double dvk = (1.-(arma::accu(delta_v(notk))>0.) )*(ln1muvk-lnuvk>lnb_m_lna(k));//imposes structured sparsity
+                double dvk = (ln1muvk-lnuvk>lnb_m_lna(k));
+                if(dvk==1){
+                    double bvktemp = R::rnorm(mu_v(k),sqrt(sigv2));
+                    bforw2 += pow(bvktemp,2);
+                    aforw2 += 1.;
+                    Gamma(vv,k) = bvktemp;
+                }
+                else{
+                    Gamma(vv,k) = 0.;
+                }
+                deltas(vv,k)=dvk;
+            }
+        }
+    }
+
+    //update w2
+    double uw = R::runif(0,1);
+    //qgamma is parameterized as shape, scale rather than shape and rate
+    double w2new = R::qgamma(uw,aforw2/2.0+1.,1.0/(.5*bforw2+1.),1,0);
+    //update nu
+    double u_nu=R::runif(0,1);
+    double nu_new = R::qbeta(u_nu,aforw2+1.,(V-1.)*K-aforw2+bnu,1,0);
+
+    return Rcpp::List::create(Rcpp::Named("w2new",w2new),
+                              Rcpp::Named("nu_new",nu_new)
+    );
+}
+
+// [[Rcpp::export]]
+arma::mat OddsRatio(unsigned int N, unsigned int J, const arma::mat &Yt) {
+    arma::mat M2_temp = arma::zeros<arma::mat>(J, J);
+    for (unsigned int j1 = 0; j1 < J - 1; ++j1) {
+        for (unsigned int j2 = j1 + 1; j2 < J; ++j2) {
+            double n11 = arma::accu(Yt.col(j1) % Yt.col(j2));
+            double n00 = arma::accu((1. - Yt.col(j1)) % (1. - Yt.col(j2)));
+            double n10 = arma::accu(Yt.col(j1) % (1. - Yt.col(j2)));
+            double n01 = N - n11 - n00 - n10;
+            M2_temp(j1, j2) = (n11 * n00) / (n10 * n01);
+        }
+    }
+    return M2_temp;
+}
+
+// [[Rcpp::export]]
+arma::mat uppertri_matrix_logical_gt(unsigned int J,const arma::mat& A,
```

Comment:
Simplify the function by using [`trimatu()`](http://arma.sourceforge.net/docs.html#trimat) view:

```cpp
# Omit the upper diagonal
AA = trimatu(A > B, 1);
```

_Note:_ Test this.